src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -0,0 +1,1011 @@
+            public bool EndStreamFlag => (Flags & FrameFlags.EndStream) != 0;
+            public bool PriorityFlag => (Flags & FrameFlags.Priority) != 0;
+
+            public static FrameHeader Read(ReadOnlySpan<byte> buffer)
+            {
+                Debug.Assert(buffer.Length >= Size);
+
+                return new FrameHeader(
+                    (buffer[0] << 16) | (buffer[1] << 8) | buffer[2],
+                    (FrameType)buffer[3],
+                    (FrameFlags)buffer[4],
+                    (int)((uint)((buffer[5] << 24) | (buffer[6] << 16) | (buffer[7] << 8) | buffer[8]) & 0x7FFFFFFF));
+            }
+
+            public void Write(Span<byte> buffer)
+            {
+                Debug.Assert(buffer.Length >= Size);
+
+                buffer[0] = (byte)((Length & 0x00FF0000) >> 16);
+                buffer[1] = (byte)((Length & 0x0000FF00) >> 8);
+                buffer[2] = (byte)(Length & 0x000000FF);
+
+                buffer[3] = (byte)Type;
+                buffer[4] = (byte)Flags;
+
+                buffer[5] = (byte)((StreamId & 0xFF000000) >> 24);
+                buffer[6] = (byte)((StreamId & 0x00FF0000) >> 16);
+                buffer[7] = (byte)((StreamId & 0x0000FF00) >> 8);
+                buffer[8] = (byte)(StreamId & 0x000000FF);
+            }
+        }
+
+        [Flags]
+        private enum FrameFlags : byte
+        {
+            None = 0,
+            
+            // Some frame types define bits differently.  Define them all here for simplicity.
+
+            EndStream =     0b00000001,
+            Ack =           0b00000001,
+            EndHeaders =    0b00000100,
+            Padded =        0b00001000,
+            Priority =      0b00100000,
+        }
+
+        private async Task SendFramesAsync(Memory<byte> frame)
+        {
+            // This can be called simultaneously by multiple Http2Streams sending their headers/request body,
+            // or by the connection itself to send connection level frames like SETTINGS.
+            // Serialize the writes.
+            await _writerLock.WaitAsync().ConfigureAwait(false);
+
+            try
+            {
+                await _stream.WriteAsync(frame).ConfigureAwait(false);
+            }
+            finally
+            {
+                _writerLock.Release();
+            }
+        }
+
+        // Note that this is safe to be called concurrently by multiple threads.
+
+        public sealed override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, bool doRequestAuth, CancellationToken cancellationToken)
+        {
+            int streamId = GetNewStreamId();
+            if (streamId == -1)
+            {
+                // TODO: Gracefully shutdown connection and mark as not usable.
+                // TODO: In debug builds, we should set the stream limit reasonably low (4K maybe?) so that we can hit it during stress testing.
+                throw new Exception("out of stream IDs");
```

Comment:
The fact that this exists at all is a bit disturbing.  In what situation would we hit this?  Only when there are 2B concurrent requests, in which case, ok, or something else?

Reply:
Yes, we'll hit this if we issue 2B requests.  From the RFC:

> Long-lived connections can
   result in an endpoint exhausting the available range of stream
   identifiers.  A client that is unable to establish a new stream
   identifier can establish a new connection for new streams.